### PR TITLE
Generate tokens for Slurmrestd using python-jose

### PR DESCRIPTION
**Pull request by matheushent**
This PR applies a big change in the code which has potential to break current versions.

Currently, tokens for Slurmrestd are generated by using subprocess and then calling `scontrol token`. This PR changes it expecting either the JWT secret key or the JWT secret key path and then generating the token by python-jose. There are two main variables expected in the dotenv file:

1. SLURMRESTD_JWT_KEY_PATH
2. SLURMRESTD_JWT_KEY_STRING

Both are optional but at least one of them must be supplied. If both are supplied, the code will prioritize the key string.

The breakthrough this PR applies is the fact that the cluster agent will no longer need to be in the same node as the slurmrestd.

### CODE REVIEW

File: cluster_agent/identity/slurmrestd.py [hunk 1]
```diff
@@ -31,15 +31,11 @@ def _load_token_from_cache(username: str) -> typing.Union[str, None]:
...
-            f"Couldn't load token from cache file {token_path}. Will acquire a new one"
-        )
+        logger.warning(f"Couldn't load token from cache file {token_path}. Will acquire a new one")
         return None
 
     try:
-        jwt.decode(
-            token, options=dict(verify_signature=False, verify_exp=True), leeway=-10
-        )
+        jwt.decode(token, options=dict(verify_signature=False, verify_exp=True), leeway=-10)
```
Comment (dusktreader) [inline cluster_agent/identity/slurmrestd.py:38]: Can we use jose to decode the tokens as well? It should be very straightforward and would eliminate some redundancy.

File: cluster_agent/identity/slurmrestd.py [hunk 2]
```diff
@@ -76,21 +72,24 @@ def acquire_token(username: str) -> str:
...
-            proc.returncode == 0, stderr.decode().strip()
-        )
-        token = stdout.decode().strip().split("=")[1]
+        logger.debug("Attempting to generate token for Slurmrestd")
+        if SETTINGS.SLURMRESTD_USE_KEY_PATH:
+            secret_key = open(SETTINGS.SLURMRESTD_JWT_KEY_PATH, "r").read()
+        else:
+            secret_key = SETTINGS.SLURMRESTD_JWT_KEY_STRING
+
+        now = datetime.now()
```
Comment (dusktreader) [inline cluster_agent/identity/slurmrestd.py:81]: Need to be careful here. By default with `datetime` you will get a timestamp in your local timezone. It might be a better idea to use UTC timestamps in all cases to eliminate ambiguity.

File: cluster_agent/settings.py [hunk 3]
```diff
@@ -63,6 +69,21 @@ def compute_extra_settings(cls, values):
         if values["SINGLE_USER_SUBMITTER"] is None:
             values["SINGLE_USER_SUBMITTER"] = values["X_SLURM_USER_NAME"]
 
+        assert any(
```
Comment (dusktreader) [inline cluster_agent/settings.py:72]: Better to raise an explicit exception here instead of relying on assert.

File: cluster_agent/settings.py [hunk 4]
```diff
@@ -63,6 +69,21 @@ def compute_extra_settings(cls, values):
...
+            [
+                values["SLURMRESTD_JWT_KEY_PATH"],
+                values["SLURMRESTD_JWT_KEY_STRING"],
+            ]
+        ), "Either SLURMRESTD_JWT_KEY_PATH or SLURMRESTD_JWT_KEY_STRING must be configured"
+
+        if not all(
```
Comment (dusktreader) [inline cluster_agent/settings.py:79]: What happens if they are both defined?

File: tests/conftest.py [hunk 5]
```diff
@@ -43,6 +43,12 @@ def mock_slurmrestd_api_cache_dir(tmp_path):
         yield _cache_dir
 
 
+@pytest.fixture(autouse=True)
+def slurmrestd_jwt_key_path(tmp_path):
+    _jwt_dir = tmp_path / "jwt.key"
```
Comment (dusktreader) [inline tests/conftest.py:48]: Do you think you should call `mkdir()` on this path here?

File: tests/identity/test_slurmrestd_identity.py [hunk 6]
```diff
@@ -148,44 +149,69 @@ def test_acquire_token__gets_a_token_from_slurm_if_one_is_not_in_the_cache(
...
     assert not token_path.exists()
 
-    process_mock = mock.Mock()
-    process_mock.communicate.return_value = (
-        bytes("SLURM_JWT=dummy-token\n".encode("utf-8")),
-        bytes(str().encode("utf-8")),
+    now = datetime.now()
+    username = "dummy-user"
+
+    mocked_datetime.now = mock.Mock()
```
Comment (dusktreader) [inline tests/identity/test_slurmrestd_identity.py:155]: Oh, this can be a fraught approach! I recommend using a library designed for this such as:
* https://pypi.org/project/time-machine/
* https://github.com/spulec/freezegun

**Review by dusktreader — CHANGES_REQUESTED** [review]
Good start! Left some comments for you.

File: cluster_agent/identity/slurmrestd.py (hunk 2 above)
Comment (matheushent) [reply cluster_agent/identity/slurmrestd.py:81]: hmm, I don't think the timezone here has any practical effect since the agent runs in customers' clusters

File: cluster_agent/settings.py (hunk 3 above)
Comment (matheushent) [reply cluster_agent/settings.py:72]: I was looking to the built-in exceptions and none of them make sense, I'd need to create another specifically for that. It's common in pydantic docs to use AssertionError here so that's why I decided for it

File: cluster_agent/settings.py (hunk 4 above)
Comment (matheushent) [reply cluster_agent/settings.py:79]: I let it briefly described in the PR description. The idea is to support both cases. In case both are defined then the code will prioritize the use of the key string directly and completely ignore the key path

File: tests/conftest.py (hunk 5 above)
Comment (matheushent) [reply tests/conftest.py:48]: hmm good point. I create the dir where the fixture is used instead in the fixture itself. I guess it makes sense to change. I shall do it

File: tests/identity/test_slurmrestd_identity.py (hunk 6 above)
Comment (matheushent) [reply tests/identity/test_slurmrestd_identity.py:155]: what do you mean by fraught approach? The code won't run in freezed time so I think freezing the time will apply an unreal behaviour

File: tests/conftest.py (hunk 5 above)
Comment (matheushent) [reply tests/conftest.py:48]: oh, I just noticed, I decided to create the dir in the test itself because the secret key string is defined in the test and I needed that to write in the file

File: cluster_agent/identity/slurmrestd.py (hunk 1 above)
Comment (matheushent) [reply cluster_agent/identity/slurmrestd.py:38]: actually, I will apply this change in this PR

File: tests/conftest.py (hunk 5 above)
Comment (matheushent) [reply tests/conftest.py:53]: resolving since I changed

File: cluster_agent/identity/slurmrestd.py (hunk 2 above)
Comment (dusktreader) [reply cluster_agent/identity/slurmrestd.py:96]: The issue I could see happening is if the cluster-agent runs in a different timezone than the slurm rest api. I just bring it up, because I have seen a lot of issues arise from the use of time zone naïve timestamps in python.

If you like, you can disregard this comment. Just know that if it ever ends up biting us, I will give you a bunch of hell about it! 😆 😆

Comment (matheushent) [reply cluster_agent/identity/slurmrestd.py:96]: this is a goog point indeed. Let's discuss it.

So, is there any way to set the timezone which slurmrestd will use? @jamesbeedy do you have any take on this?

Comment (matheushent) [reply cluster_agent/identity/slurmrestd.py:96]: I couldn't find any information about that on the [docs](https://slurm.schedmd.com/slurmrestd.html)

File: cluster_agent/settings.py (hunk 3 above)
Comment (dusktreader) [reply cluster_agent/settings.py:72]: I think it's best practice to not use assertions to check conditions in production code because assertions can be disabled depending on the execution mode of python.

You could use pydantic's `ValidationError` as the exception type. I think that would fit very well.

File: cluster_agent/settings.py (hunk 4 above)
Comment (dusktreader) [reply cluster_agent/settings.py:79]: I am uncomfortable with that implicit behavior. I think it should explicitly require one or the other so there is no chance for ambiguity.

**Comment by dusktreader** [comment]
Task linked: [ARMADA-438 Discovery: generate tokens for Slurm REST API auth](https://app.clickup.com/t/18022949/ARMADA-438)

File: tests/identity/test_slurmrestd_identity.py (hunk 6 above)
Comment (dusktreader) [reply tests/identity/test_slurmrestd_identity.py:183]: I mean that relying on a mock to manipulate the current time is dangerous. There are lots of edge cases and odd behavior.

When testing against code that depends on getting the current time, the best approach is to use a proven mechanism for setting the current time explicitly. It is the safest and most reliable way of testing time critical code.

It is true that the code does not run in frozen time, however, a unit testing environment is artificial and we need to be able to control the environment. Freezing and manipulating the wall-clock time is a tried and true approach.

For example, the Armasec unit tests use this when dealing with token expiration logic: https://github.com/omnivector-solutions/armasec/blob/main/tests/test_token_manager.py#L68

Comment (matheushent) [reply tests/identity/test_slurmrestd_identity.py:183]: You have a fair point.

Looking at [freezegun's source code](https://github.com/spulec/freezegun/blob/ac4d92ec6d3545b113aba96dc444fd055219af9b/freezegun/api.py#L816) we can see they basically also mock the time with a different approach. I cannot see the difference that can lead us in edge cases. Can you elaborate a little bit more the edge cases you see? It could help me to develop new tests as well.

I think having it not frozen enables us to know if the code can run at any time, which indeed it is supposed to.

File: cluster_agent/settings.py (hunk 3 above)
Comment (matheushent) [reply cluster_agent/settings.py:72]: I looked at the `ValidationError` class first and this class requires a pydantic model as parameter. Not sure if it is useful in this context. So, I went to [pydantic's docs](https://pydantic-docs.helpmanual.io/usage/validators/) and found examples with assertion. Indeed, not my favourite but I'd guess it is the recommended. Does it make sense for you?

Also, to disable the assertion error the python code must run with the `-O` flag, which isn't the case here because the agent is supposed to run with the `agentrun` entrypoint defined in the [setup.py](https://github.com/omnivector-solutions/cluster-agent/blob/main/setup.py#L55) file, and indeed, [is what happens in our charm](https://github.com/omnivector-solutions/charm-cluster-agent/blob/main/src/templates/ovs-cluster-agent.service#L10).

File: cluster_agent/settings.py (hunk 4 above)
Comment (matheushent) [reply cluster_agent/settings.py:79]: let me figure it out the changes I will need to make in order to change it

File: cluster_agent/settings.py (hunk 3 above)
Comment (fschuch) [reply cluster_agent/settings.py:72]: I saw this comment I came to see because I'm using assert in one of my drafts. A way I see to go around this is using [check-expressions](https://py-buzz.readthedocs.io/en/latest/features.html#check-expressions), from py-buzz.

Comment (matheushent) [reply cluster_agent/settings.py:72]: it would be the same as `if condition: raise Error`, which is basically what the AssertionError here is doing in a different way

File: tests/identity/test_slurmrestd_identity.py (hunk 6 above)
Comment (dusktreader) [reply tests/identity/test_slurmrestd_identity.py:183]: Here is an interesting article about the problem and how a library helps:
https://adamj.eu/tech/2020/06/03/introducing-time-machine/

Freezing time is important because you don't know when a test will be run. Sometimes your logic makes assumption about events happening in the same day (what if the test runs exactly at midnight) or same hour (what happens if the test runs exactly at the hour). When you are creating tests that involve time, it's really important that you control the time parameter. That way, you can test that it works correctly in a normal situation and also covers the odd edge-cases.

This is best practice and an important tool to use when testing around timestamps.

Comment (matheushent) [reply tests/identity/test_slurmrestd_identity.py:183]: The article is indeed very interesting, but look at his point on why mock isn't good for mocking time:

[Using unittest.mock works, but it’s often inaccurate as each patcher can only mock a single function reference. This inaccuracy is exacerbated when mocking the current time, as there are many different functions that return the current time, in different formats. Due to the way Python’s imports work, it takes a lot of mocks to replace every instance of functions from datetime and time in a code path, and is sometimes impossible.](https://adamj.eu/tech/2020/06/03/introducing-time-machine/#:~:text=Using%20unittest.mock,is%20sometimes%20impossible.)

If you think precisely on this you shall realize that the current implementation I did has nothing against Mr. Johnson's thinking.

Also, I found very weird the fact these other packages for freezing the time change the C level of python. This definitely won't simulate the environment in which the code will run. What if I change the python interpreter then? What problems could we have? If you see, every approach has its downsides.

If we want some granularity here I'd prefer to generate the datetime randomly based on a seed.

I also want other folks' thinkings because I'm stricted convinced using this other packages isn't a good practice at all.

@julianaklulo @fschuch give your thoughts please

File: cluster_agent/settings.py (hunk 4 above)
Comment (matheushent) [reply cluster_agent/settings.py:79]: so, removing the ambiguity as you say is taking longer than expected. There are a bunch of tests to refactor. Is there any potential problem you see leaving the way it is? I, with honesty, don't see any **if** well documented (which is not the case. I'd need to write the docs for that)

File: tests/identity/test_slurmrestd_identity.py (hunk 6 above)
Comment (julianaklulo) [reply tests/identity/test_slurmrestd_identity.py:183]: I've been using `freezegun` since the first time I needed to manipulate the current time in a test, because it is the recommended approach for the reasons @dusktreader already mentioned above.

I agree that mocking the `datetime` directly is an odd way of doing this, and since we've been using `pytest-freezegun` in other projects, I'd suggest we set it as a standard across all projects.

File: cluster_agent/settings.py (hunk 4 above)
Comment (matheushent) [reply cluster_agent/settings.py:84]: according to private conversations with @dusktreader, we agreed that the current implementation is fine once we have clear documentation about the behaviour and the agent logs a warning in case both settings are passed. Both commits [42691ec](https://github.com/omnivector-solutions/cluster-agent/pull/54/commits/42691eccebdcbb3abf5c3c6e207d10d7d4210053) and [74eb9bd](https://github.com/omnivector-solutions/cluster-agent/pull/54/commits/74eb9bd81590b4606b1225f01d0ccea46d617961) handle such requirements.

I'm resolving this conversation. Feel free to review again.

File: tests/identity/test_slurmrestd_identity.py (hunk 6 above)
Comment (matheushent) [reply tests/identity/test_slurmrestd_identity.py:183]: according to private conversations with @dusktreader, we agreed the current implementation is fine but he alerted with examples why freezegun or timemachine are good packages for handling timestamp manipulation during tests. Since we resolved this conversation in private I will resolve here as well.

Furthermore, you folks feel free to give your thoughts even if the conversation is resolved.

**Review by dusktreader — APPROVED** [review]

**Review by dusktreader — APPROVED** [review]
